samples/solutions/producer_consumer.py

```diff
@@ -0,0 +1,244 @@
+sample_log_handler.setFormatter(log_formatter)
+logger = logging.getLogger(__name__)
+logger.addHandler(sample_log_handler)
+
+
+def get_type_name(e):
+    return type(e).__name__
+
+
+class Application(object):
+    async def initiate(self):
+        self.connected_event = asyncio.Event()
+        self.disconnected_event = asyncio.Event()
+        self.exit_app_event = asyncio.Event()
+        self.disconnected_event.set()
+
+        self.message_queue = asyncio.Queue()
+        self.device_client = None
+        self.first_connect = True
+        # Power factor for increasing interval between consecutive connection attempts.
+        # This will increase with iteration
+        self.retry_increase_factor = 1
+        # The nth number for attempting connection
+        self.sleep_time_between_conns = ATTEMPT_INTERVAL
+        self.try_number = 1
+
+    async def create_client(self, conn_str):
+        try:
+            # Create a Device Client
+            self.device_client = IoTHubDeviceClient.create_from_connection_string(conn_str)
+            # Attach the connection state handler
+            self.device_client.on_connection_state_change = self.handle_on_connection_state_change
+        except Exception as e:
+            self.log_error_and_print(
+                "Caught exception while trying to attach handler : {}".format(get_type_name(e))
+            )
+
+    async def handle_on_connection_state_change(self):
+        self.log_info_and_print(
+            "handle_on_connection_state_change fired. Connected status : {}".format(
+                self.device_client.connected
+            )
+        )
+        if self.device_client.connected:
+            self.log_info_and_print("Connected connected_event is set...")
+            self.disconnected_event.clear()
+            self.connected_event.set()
+            self.retry_increase_factor = 1
+            self.sleep_time_between_conns = ATTEMPT_INTERVAL
+            self.try_number = 1
+
+        else:
+            self.log_info_and_print("Disconnected connected_event is set...")
+            self.disconnected_event.set()
+            self.connected_event.clear()
+
+    async def enqueue_message(self):
+        message_id = 0
+        while True:
+            message_id += 1
+            msg = Message("current wind speed ")
+            msg.message_id = message_id
+            msg.content_type = "application/json"
+            self.log_info_and_print("Created a message...")
+            self.message_queue.put_nowait(msg)
+            await asyncio.sleep(TELEMETRY_INTERVAL)
+            if self.exit_app_event.is_set():
+                return
+
+    async def wait_for_connect_and_send_telemetry(self):
+        while True:
+            if not self.device_client.connected:
+                self.log_info_and_print("waiting for connection ...")
+                await self.connected_event.wait()
+            else:
+                msg = await self.message_queue.get()
```

Comment:
>msg = await self.message_queue.get() [](http://example.com/codeflow?start=16&length=36)

`asyncio.queue.get()` will wait until an item is in the queue. What happens if `self.exit_app_event` gets set while this all is waiting for an item to come into the queue?

Reply:
changed it to first completed with the queue and exit app